{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Task 10: Labelling Relevant Regions (cont.)\n",
    "## Starting with the ACE2 TAD (chrX:15,200,000-15,800,000) and expanding the algorithm to larger regions:\n",
    "### Add labels to:\n",
    "1. Portein-coding vs. Non-protein-coding regions:\n",
    "\n",
    "Protein-coding|Non-protein-coding\n",
    ":--:|:--:\n",
    "Coding sequence of exons from protein-coding transcripts|Exons from non-protein-coding transcripts, UTRs from protein-coding transcripts\n",
    "\n",
    "### Using [hg38.ncbiRefSeqCurated.gtf](http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/):"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "./genePredToGtf -utr hg38 ncbiRefSeqCurated hg38.ncbiRefSeqCurated.gtf\n",
    "cp hg38.ncbiRefSeqCurated.gtf /mnt/c/Users/bkirs/Documents/Wasserman/Task10\n",
    "\n",
    "**To reveal the number of each of the different feature types in the gtf file:**\n",
    "cat hg38.ncbiRefSeqCurated.gtf | grep -v \"^#\" | cut -f3 | sort | uniq -c | sort -k1rn\n",
    "\n",
    "\n",
    "cat hg38.ncbiRefSeqCurated.gtf | awk 'BEGIN{FS=\"\\t\"}{split($9,a,\";\"); if($1==\"chrX\" && $4>=15200000 && $5<=15800000) print a[2]\"\\t\"$1\"\\t\"$2\"\\t\"$3\"\\t\"$4\"\\t\"$5\"\\t\"$6\"\\t\"$7\"\\t\"$8;}' | sed 's/transcript_id \"//' | sed 's/\"//g' | sed 's/ //g' | sed \"1i\\transcriptId\\tChromosome\\tSource\\tFeature\\tStart\\tEnd\\tScore\\tStrand\\tFrame\" > ncbiRefSeqCuratedId.txt\n",
    "\n",
    "sudo mysql -A -e \"CREATE DATABASE task10\"\n",
    "sudo mysql -A -e \"USE task10\"\n",
    "\n",
    "sudo mysql -A -e \"CREATE TABLE ncbiRefSeqGTF (ncbiId VARCHAR(20), seqname VARCHAR(5), source VARCHAR(10), feature VARCHAR(20), start VARCHAR(20), end VARCHAR(20), score VARCHAR(5), strand CHAR(1), frame CHAR(1))\" task10\n",
    "sudo mysql -A -e \"LOAD DATA LOCAL INFILE 'ncbiRefSeqCuratedId.txt' INTO TABLE ncbiRefSeqGTF IGNORE 1 LINES\" task10\n",
    "sudo mysql -A -e \"DELETE FROM ncbiRefSeqGTF WHERE ncbiId LIKE 'X%'\" task10\n",
    "sudo mysql -A -e \"DELETE FROM ncbiRefSeqGTF WHERE ncbiId LIKE 'Y%'\" task10\n",
    "sudo mysql -A -e \"DELETE FROM ncbiRefSeqGTF WHERE ncbiId LIKE 'NM%' AND FEATURE LIKE 'exon'\" task10\n",
    "\n",
    "sudo mysql -A -e \"CREATE TABLE codingLabels (feature VARCHAR(20), status VARCHAR(20), scoreVal VARCHAR(5))\" task10\n",
    "sudo mysql -A -e \"LOAD DATA LOCAL INFILE 'codingLabels.txt' INTO TABLE codingLabels IGNORE 1 LINES\" task10\n",
    "\n",
    "sudo mysql -A -e \"CREATE TABLE allInfo (ncbiId VARCHAR(20), name VARCHAR(20), sequenceType VARCHAR(20), chrom VARCHAR(5), chromStart VARCHAR(20), chromEnd VARCHAR(20), strand CHAR(1), status VARCHAR(20), scoreVal VARCHAR(5))\" task10\n",
    "sudo mysql -A -e \"SELECT a.ncbiId, name, a.feature, seqname, start, end, strand, status, scoreVal FROM ncbiRefSeqGTF a INNER JOIN test.ncbiRefSeqLinkSelect b ON a.ncbiId = b.ncbiId INNER JOIN codingLabels c ON a.feature = c.feature\" task10 > allInfo.txt\n",
    "sudo mysql -A -e \"LOAD DATA LOCAL INFILE 'allInfo.txt' INTO TABLE allInfo IGNORE 1 LINES\" task10\n",
    "\n",
    "sudo mysql -A -e \"SELECT chrom, chromStart, chromEnd, name, scoreVal, strand FROM allInfo GROUP BY chromStart, chromEnd\" task10 > codingTAD.txt\n",
    "cat codingTAD.txt | sed '1d' > codingTAD.bed\n",
    "\n",
    "uniq codingTAD.bed | awk '{$1=$1\"_\"$5; if($5!=500) print}' | sort -k1,1 -k2,2n > codingTAD.group2.bed\n",
    "bedops --range 0:1 --merge codingTAD.group2.bed > codingTAD.merge.bed\n",
    "awk '{VAR=$1; split(VAR, arr, \"_\"); $1=arr[1]; $5=arr[2]; if($5==1000) $4=\"coding\"; else $4=\"non-coding\"; print}' codingTAD.merge.bed > UCSC.coding.bed\n",
    "\n",
    "sed 's/ \\+/\\t/g' UCSC.coding.bed > UCSC.coding.txt\n",
    "sudo mysql -A -e \"CREATE TABLE codingUCSC (chrom VARCHAR(5), chromStart VARCHAR(20), chromEnd VARCHAR(20), name VARCHAR(20), score VARCHAR(5))\" task10\n",
    "sudo mysql -A -e \"LOAD DATA LOCAL INFILE 'UCSC.coding.txt' INTO TABLE codingUCSC\" task10\n",
    "sudo mysql -A -e \"ALTER TABLE codingUCSC ADD length VARCHAR(5) GENERATED ALWAYS AS (chromEnd-chromStart) STORED\" task10\n",
    "sudo mysql -A -e \"SELECT SUM(length)/15200000*100 FROM codingUCSC WHERE name LIKE 'coding'\" task10\n",
    "sudo mysql -A -e \"SELECT SUM(length)/(15800000-15200000)*100 FROM codingUCSC WHERE name LIKE 'coding'\" task10\n",
    "sudo mysql -A -e \"SELECT SUM(length)/(15800000-15200000)*100 FROM codingUCSC WHERE name LIKE 'non-coding'\" task10"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sudo mysql -A -e \"SELECT chrom, chromStart, chromEnd, CONCAT(ncbiId,'_',chrom,'_',chromStart,'_',chromEnd,'_',status) FROM allInfo GROUP BY chromStart, chromEnd\" task10 > bedtoolsInfo.txt\n",
    "sudo mysql -A -e \"SELECT chrom, chromStart, chromEnd, CONCAT(ncbiId,'_',chrom,'_',chromStart,'_',chromEnd,'_',status) FROM allInfo WHERE status LIKE 'protein-coding' GROUP BY chromStart, chromEnd\" task10 > bedtoolsProtein.txt\n",
    "sudo mysql -A -e \"SELECT chrom, chromStart, chromEnd, CONCAT(ncbiId,'_',chrom,'_',chromStart,'_',chromEnd,'_',status) FROM allInfo WHERE status LIKE 'non-protein-coding' GROUP BY chromStart, chromEnd\" task10 > bedtoolsNonProtein.txt\n",
    "\n",
    "uniq bedtoolsSNonProtein.txt | sort -k1,1 -k2,2n | bedtools merge -i stdin -c 4 -o collapse | awk '{$5=\"non-protein-coding\"; print}' > nonproteinMerge.bed\n",
    "uniq bedtoolsProtein.txt | sort -k1,1 -k2,2n | bedtools merge -i stdin -c 4 -o collapse | awk '{$5=\"protein-coding\"; print}' > proteinMerge.bed\n",
    "cat nonProteinMerge.bed proteinMerge.bed > codingTAD.merge2.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. Immediate promoters:  \n",
    "### Using FANTOM5 TSS peaks:\n",
    "**TSS Peaks and Enhancers:**  \n",
    "> - TSS (CAGE) peaks are used as anchor to define promoters and as unit of promoter-level expression analysis: Two subsets of the peaks are defined based on evidence of read counts, depending on scopes of subsequent analyses.  \n",
    ">> - The first subset (referred as robust set of the peaks) is named as \"p#@GENE_SYMBOL\" if associated with 5'-end of known genes, or \"p@CHROM:START..END,STRAND\" otherwise.  "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sort -k1,1 -k2,2n FANTOM5.TSSpeaks.hg38.txt > promoters.sort.bed\n",
    "bedops --merge promoters.sort.bed | awk '{$4=$6=\".\"; $5=0; $7=$2; $8=$3; $9=\"255,0,0\"; print}' > UCSC.promoters.bed\n",
    "\n",
    "\n",
    "sudo mysql -A -e \"CREATE TABLE promotersUCSC (chrom VARCHAR(5), chromStart VARCHAR(20), chromEnd VARCHAR(20))\" task10\n",
    "sed 's/ \\+/\\t/g' UCSC.promoters.bed | cut -f 1-3 > UCSC.promoters.txt\n",
    "sudo mysql -A -e \"LOAD DATA LOCAL INFILE 'UCSC.promoters.txt' INTO TABLE promotersUCSC\" task10\n",
    "sudo mysql -A -e \"ALTER TABLE promotersUCSC ADD length VARCHAR(5) GENERATED ALWAYS AS (chromEnd-chromStart) STORED\" task10\n",
    "sudo mysql -A -e \"SELECT SUM(length)/(15800000-15200000)*100 FROM promotersUCSC\" task10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Intronic vs. Intergenic:\n",
    "### Using UCSC Table Browser (ncbiRefSeqCurated):\n",
    "**Reference for degining genomic regions:** https://davetang.org/muse/2013/01/18/defining-genomic-regions/"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "awk '{VAR=$4; split(VAR, arr, \"_\"); $1=arr[1]\"_\"arr[2]; $4=\"intron\"; print $1, $2+1, $3, $4}' hg38_NCBIRefSeqCurated_intron.bed | sed 's/ \\+/\\t/g' > introns_TAD.txt\n",
    "sudo mysql -A -e \"CREATE TABLE ncbiRefSeqIntrons (id VARCHAR(20), chromStart VARCHAR(20), chromEnd VARCHAR(20), feature VARCHAR(10))\" task10\n",
    "sudo mysql -A -e \"LOAD DATA LOCAL INFILE 'introns_TAD.txt' INTO TABLE ncbiRefSeqIntrons\" task10\n",
    "\n",
    "sed 's/ \\+/\\t/g' codingTAD.merge2.bed > exons_TAD.txt\n",
    "sudo mysql -A -e \"CREATE TABLE ncbiRefSeqExons (chrom VARCHAR(5), chromStart VARCHAR(20), chromEnd VARCHAR(20), parents VARCHAR(100), type VARCHAR(20))\" task10\n",
    "sudo mysql -A -e \"LOAD DATA LOCAL INFILE 'exons_TAD.txt' INTO TABLE ncbiRefSeqExons\" task10\n",
    "\n",
    "sudo mysql -A -e \"CREATE TABLE regionsRefSeq (chrom VARCHAR(5), chromStart VARCHAR(20), chromEnd VARCHAR(20))\" task10\n",
    "\n",
    "sort -k1,1 -k2,2n hg38_NCBIRefSeqCurated_intron.bed | bedtools merge -i stdin -c 2,3,4 -o max,min,collapse > intronTAD.intersect.bed\n",
    "awk '$4 < $5' intronTAD.intersect.bed | awk '{print $1, $4, $5, $6, $7=\"intron\"}' | sed 's/ \\+/\\t/g' > introns_TAD2.txt"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Obtain all \"exonic\" (i.e. protein-coding and non-protein-coding) regions:\n",
    "\n",
    "Option 1: From UCSC Table Browser exon files:\n",
    "\n",
    "cat hg38_NCBIRefSeqCurated_CDS.bed hg38_NCBIRefSeqCurated_3UTR.bed hg38_NCBIRefSeqCurated_5UTR.bed | sort -k1,1 -k2,2n | bedtools merge -i stdin -c -4 -o distinct\n",
    "\n",
    "*chromStart is off-by-one*\n",
    "\n",
    "Option 2: From combined exon file obtained in the protein-coding vs. non-protein-coding stage:\n",
    "\n",
    "sort -k1,1 -k2,2n exons_TAD.txt | bedtools merge -i stdin -d 1 -c 4 -o distinct\n",
    "\n",
    "*both commands produce the same output when the off-by-one factor is corrected, but the name column uses different id conventions*"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "To define intronic regions, we just need to subtract the exonic region from the genic region. \n",
    "\n",
    "Within regions that are classified as a \"transcript,\" \"intronic regions\" are made up by all features that were not previously labelled as \"exonic\": This ensures that if a region was ever observed to have at least one exonic feature, it recieves the \"exonic\" classification rather than the \"intronic\" label.\n",
    "\n",
    "sort -k1,1 -k2,2n exons_TAD.txt | bedtools merge -i stdin -d 1 -c 4 -o distinct > exons_TAD2.txt\n",
    "\n",
    "sudo mysql -A -e \"SELECT * FROM ncbiRefSeqGTF where feature like 'transcript'\" task10 > transcripts_TAD.txt\n",
    "awk '{print $2, $5, $6, $1, $4}' transcripts_TAD.txt | sed 's/ \\+/\\t/g' | sed '1d' > transcripts_TAD2.txt\n",
    "\n",
    "bedtools subtract -a transcripts_TAD2.txt -b exons_TAD2.txt | sort -k1,1 -k2,2n | bedtools merge -i stdin -c 4 -o distinct | awk '$5=\"intron\"' | sed 's/ \\+/\\t/g' > introns_TAD3.txt\n",
    "\n",
    "To define intergenic regions, we use complementBed to find regions not covered by genes.\n",
    "\n",
    "Within the region of interest (i.e. the ACE2 TAD, in this case), \"intergenic regions\" are made up by all features that were not previously labelled as \"transcripts\": This ensures that if a region was ever observed to be spanned by at least one transcipt, it recieves a transcript-associated classification (i.e. protein-coding, non-protein-coding, or intron) rather than the \"intergenic\" label.\n",
    "\n",
    "*the \"genome file\" contains a single line: chrX 15200000 15800000 ACE2.TAD*\n",
    "bedtools subtract -a genome_TAD.txt -b transcripts_TAD2.txt | awk '$4=\"intergene\"' | sed\n",
    " 's/ \\+/\\t/g' > intergene_TAD.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exon reading frames: https://groups.google.com/a/soe.ucsc.edu/g/genome/c/U-w4b_ZS2j0?pli=1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Repetitive Sequences:\n",
    "### Using [RepeatMasker](http://www.repeatmasker.org/faq.html):\n",
    "**How to read the results:** http://www.repeatmasker.org/webrepeatmaskerhelp.html  \n",
    "\n",
    "> - Classes: 'SINE', 'LINE', 'LTR', 'DNA', 'Satellite', anything with 'RNA' in it, 'Simple_repeat', and 'Other' or 'Unknown  "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "1. Obtain all repeats in the desired region:\n",
    "\n",
    "mysql -h genome-mysql.soe.ucsc.edu -ugenome -A -e \"select genoName, genoStart, genoEnd, strand, repName, repClass, repFamily, swScore from rmsk where genoName like 'chrX' and genoStart >= 15200000 and genoEnd <= 15800000\" hg38 > hg38.rmsk.txt\n",
    "\n",
    "2. Obtain only LINEs and SINEs:\n",
    "\n",
    "mysql -h genome-mysql.soe.ucsc.edu -ugenome -A -e \"select genoName, genoStart, genoEnd, strand, repName, repClass, repFamily, swScore from rmsk where genoName like 'chrX' and genoStart >= 15200000 and genoEnd <= 15800000 and (repClass like 'SINE' or repClass like 'LINE')\" hg38 > hg38.rmsk.LINES_SINES.txt\n",
    "\n",
    "3. Obtain all possible families of LINEs and SINEs:\n",
    "\n",
    "mysql -h genome-mysql.soe.ucsc.edu -ugenome -A -e \"select distinct repFamily from rmsk where repClass like 'SINE' or repClass like 'LINE'\" hg38"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating Tracks For All Classification Systems:"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "./genePredToGtf -utr hg38 ncbiRefSeqCurated hg38.ncbiRefSeqCurated.gtf #Download and convert source data\n",
    "\n",
    "cat hg38.ncbiRefSeqCurated.gtf | awk -F \"\\t\" 'BEGIN{OFS=\"\\t\";} ($1==\"chrX\" && (($5 >= 15200000 && $5 <= 15800000) || ($4 >= 15200000 && $4 <= 15800000))) && ($3==\"exon\") {print $1,$4,$5}' | bedtools sort | bedtools merge -i stdin | gzip > exon.bed.gz #Extract exonic regions\n",
    "\n",
    "cat hg38.ncbiRefSeqCurated.gtf | awk -F \"\\t\" 'BEGIN{OFS=\"\\t\";} ($1==\"chrX\" && (($5 >= 15200000 && $5 <= 15800000) || ($4 >= 15200000 && $4 <= 15800000))) && ($3==\"transcript\") {print $1,$4,$5}' | bedtools sort | bedtools subtract -a stdin -b exon.bed.gz | gzip > intron.bed.gz  #Extract intronic regions\n",
    "\n",
    "mysql --user=genome --host=genome-mysql.cse.ucsc.edu -A -B -N -e \"SELECT chrom, size FROM hg38.chromInfo WHERE chrom LIKE 'chrX'\" > hg38.genome #Create genome file\n",
    "\n",
    "cat hg38.ncbiRefSeqCurated.gtf | awk -F \"\\t\" 'BEGIN{OFS=\"\\t\";} ($1==\"chrX\" && (($5 >= 15200000 && $5 <= 15800000) || ($4 >= 15200000 && $4 <= 15800000))) && ($3==\"transcript\") {print $1,$4,$5}' | bedtools sort | bedtools complement -i stdin -g hg38.genome | awk 'BEGIN{OFS=\"\\t\"}; {if($2<15200000) $2=15200000}; {if($3>15800000) $3=15800000}; $3>=15200000 && $2 <=15800000' | gzip > intergenic.bed.gz #Extract intergenic regions\n",
    "\n",
    "**change to:**\n",
    "./fetchChromSizes hg38 > hg38.chrom.sizes\n",
    "cat hg38.ncbiRefSeqCurated.gtf | awk -F \"\\t\" 'BEGIN{OFS=\"\\t\";} ($1==\"chrX\" && (($5 >= 15200000 && $5 <= 15800000) || ($4 >= 15200000 && $4 <= 15800000))) && ($3==\"transcript\") {print $1,$4,$5}' | bedtools sort | bedtools complement -i stdin -g hg38.chrom.sizes | awk 'BEGIN{OFS=\"\\t\"}; {if($2<15200000) $2=15200000}; {if($3>15800000) $3=15800000}; $1==\"chrX\" && $3>=15200000 && $2 <=15800000'\n",
    "\n",
    "\n",
    "cat hg38.ncbiRefSeqCurated.gtf | awk -F \"\\t\" 'BEGIN{OFS=\"\\t\";} ($1==\"chrX\" && (($5 >= 15200000 && $5 <= 15800000) || ($4 >= 15200000 && $4 <= 15800000))) && ($3==\"CDS\" || $3==\"start_codon\" || $3==\"stop_codon\")' |  bedtools sort | bedtools merge -i stdin | gzip > coding.bed.gz #Extract protein-coding exonic regions\n",
    "\n",
    "cat hg38.ncbiRefSeqCurated.gtf | awk -F \"\\t\" 'BEGIN{OFS=\"\\t\";} ($1==\"chrX\" && (($5 >= 15200000 && $5 <= 15800000) || ($4 >= 15200000 && $4 <= 15800000))) && (($3~/UTR/) || ($9~/NR/ && $3==\"exon\")) {print $1,$4,$5}' | bedtools sort | bedtools subtract -a stdin -b ncbiRefSeqCurated_coding.bed.gz | sort -k1,1 -k2,2n | bedtools merge -i stdin | gzip > noncoding.bed.gz #Extract non-protein-coding exonic regions\n",
    "\n",
    "cat hg38.FANTOM5_TSSpeaks.bed | bedtools sort | bedtools merge -i stdin | awk '{print $1, $2, $3}' | gzip > promoter.bed.gz #Use Table Browser FANTOM5 Hub to extract promoters: Filter for genomic region during the download step.\n",
    "\n",
    "mysql -h genome-mysql.soe.ucsc.edu -ugenome -A -B -N -e \"SELECT genoName, genoStart, genoEnd FROM rmsk WHERE genoName LIKE 'chrX' AND genoStart >= 15200000 AND genoEnd <= 15800000 AND repClass LIKE  'LINE'\" hg38 | bedtools sort | bedtools merge -i stdin | gzip > LINE.bed.gz #Extract all LINEs\n",
    "\n",
    "mysql -h genome-mysql.soe.ucsc.edu -ugenome -A -B -N -e \"SELECT genoName, genoStart, genoEnd FROM rmsk WHERE genoName LIKE 'chrX' AND genoStart >= 15200000 AND genoEnd <= 15800000 AND repClass LIKE 'SINE' AND repFamily NOT LIKE 'Alu'\" hg38 | bedt\n",
    "ools sort | bedtools merge -i stdin | gzip > SINE.bed.gz  #Extract all non-Alu SINEs\n",
    "\n",
    "mysql -h genome-mysql.soe.ucsc.edu -ugenome -A -B -N -e \"SELECT genoName, genoStart, genoEnd FROM rmsk WHERE genoName LIKE 'chrX' AND genoStart >= 15200000 AND genoEnd <= 15800000 AND repFamily LIKE 'Alu'\" hg38 | bedtools sort | bedtools merge -i\n",
    " stdin | gzip > Alu.bed.gz  #Extract all Alus"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "cat hg38.ncbiRefSeqCurated.gtf | awk -F \"\\t\" 'BEGIN{OFS=\"\\t\";} ($1==\"chrX\" && (($5 >= 15200000 && $5 <= 15800000) || ($4 >= 15200000 && $4 <= 15800000))) && ($3==\"exon\") {print $1,$4,$5}' | bedtools sort | bedtools merge -i stdin | awk 'BEGIN{OFS=\"\\t\";}; {if($2<15200000) $2=15200000}; {if($3>15800000) $3=15800000}; {print $1,$2,$3,\"exon\",0,\".\",$2,$3,\"0,255,0\"}' | uniq | gzip > exon.bed.gz\n",
    "\n",
    "cat hg38.ncbiRefSeqCurated.gtf | awk -F \"\\t\" 'BEGIN{OFS=\"\\t\";} ($1==\"chrX\" && (($5 >= 15200000 && $5 <= 15800000) || ($4 >= 15200000 && $4 <= 15800000))) && ($3==\"transcript\") {print $1,$4,$5}' | bedtools sort | bedtools merge -i stdin | bedtools subtract -a stdin -b exon.bed.gz | awk 'BEGIN{OFS=\"\\t\";}; {if($2<15200000) $2=15200000}; {if($3>15800000) $3=15800000}; {print $1,$2,$3,\"intron\",0,\".\",$2,$3,\"255,0,0\"}' | gzip > intron.bed.gz\n",
    "\n",
    "cat hg38.ncbiRefSeqCurated.gtf | awk -F \"\\t\" 'BEGIN{OFS=\"\\t\";} ($1==\"chrX\" && (($5 >= 15200000 && $5 <= 15800000) || ($4 >= 15200000 && $4 <= 15800000))) && ($3==\"transcript\") {print $1,$4,$5}' | bedtools sort | bedtools complement -i stdin -g hg38.genome | awk 'BEGIN{OFS=\"\\t\"}; {if($2<15200000) $2=15200000}; {if($3>15800000) $3=15800000}; $3>=15200000 && $2 <=15800000' | awk 'BEGIN{OFS=\"\\t\";} {print $1,$2,$3,\"intergenic\",0,\".\",$2,$3,\"0,0,255\"}' | gzip > intergenic.bed.gz\n",
    "\n",
    "cat hg38.ncbiRefSeqCurated.gtf | awk -F \"\\t\" 'BEGIN{OFS=\"\\t\";} ($1==\"chrX\" && (($5 >= 15200000 && $5 <= 15800000) || ($4 >= 15200000 && $4 <= 15800000))) && ($3==\"CDS\" || $3==\"start_codon\" || $3==\"stop_codon\")' |  bedtools sort | bedtools merge -i stdin | awk 'BEGIN{OFS=\"\\t\";}; {if($2<15200000) $2=15200000}; {if($3>15800000) $3=15800000}; {print $1,$2,$3,\"coding\",0,\".\",$2,$3,\"0,102,0\"}' | uniq | gzip > coding.bed.gz\n",
    "\n",
    "cat hg38.ncbiRefSeqCurated.gtf | awk -F \"\\t\" 'BEGIN{OFS=\"\\t\";} ($1==\"chrX\" && (($5 >= 15200000 && $5 <= 15800000) || ($4 >= 15200000 && $4 <= 15800000))) && (($3~/UTR/) || ($9~/NR/ && $3==\"exon\")) {print $1,$4,$5}' | bedtools sort | bedtools subtract -a stdin -b coding.bed.gz | sort -k1,1 -k2,2n | bedtools merge -i stdin | awk 'BEGIN{OFS=\"\\t\";}; {if($2<15200000) $2=15200000}; {if($3>15800000) $3=15800000}; {print $1,$2,$3,\"noncoding\",0,\".\",$2,$3,\"153,255,153\"}' | uniq | gzip > noncoding.bed.gz\n",
    "\n",
    "cat hg38.FANTOM5_TSSpeaks.bed | bedtools sort | bedtools merge -i stdin | awk 'BEGIN{OFS=\"\\t\";} {print $1,$2,$3,\"promoter\",0,\".\",$2,$3,\"127,0,255\"}' | uniq | gzip > promoter.bed.gz\n",
    "\n",
    "mysql -h genome-mysql.soe.ucsc.edu -ugenome -A -B -N -e \"SELECT genoName, genoStart, genoEnd FROM rmsk WHERE genoName LIKE 'chrX' AND ((genoEnd >= 15200000 && genoEnd <= 15800000) OR (genoStart >= 15200000 AND genoStart <= 15800000)) AND repClass LIKE  'LINE'\" hg38 | bedtools sort | bedtools merge -i stdin | awk 'BEGIN{OFS=\"\\t\";} {print $1,$2,$3,\"LINE\",0,\".\",$2,$3,\"255,255,0\"}' | uniq | gzip > LINE.bed.gz\n",
    "\n",
    "mysql -h genome-mysql.soe.ucsc.edu -ugenome -A -B -N -e \"SELECT genoName, genoStart, genoEnd FROM rmsk WHERE genoName LIKE 'chrX' AND ((genoEnd >= 15200000 && genoEnd <= 15800000) OR (genoStart >= 15200000 AND genoStart <= 15800000)) AND repClass LIKE 'SINE' AND repFamily NOT LIKE 'Alu'\" hg38 | bedtools sort | bedtools merge -i stdin | awk 'BEGIN{OFS=\"\\t\";}; {if($2<15200000) $2=15200000}; {if($3>15800000) $3=15800000}; {print $1,$2,$3,\"SINE\",0,\".\",$2,$3,\"255,204,153\"}' | uniq | gzip > SINE.bed.gz\n",
    "\n",
    "mysql -h genome-mysql.soe.ucsc.edu -ugenome -A -B -N -e \"SELECT genoName, genoStart, genoEnd FROM rmsk WHERE genoName LIKE 'chrX' AND ((genoEnd >= 15200000 && genoEnd <= 15800000) OR (genoStart >= 15200000 AND genoStart <= 15800000)) AND repFamily LIKE 'Alu'\" hg38 | bedtools sort | bedtools merge -i stdin | awk 'BEGIN{OFS=\"\\t\";}; {if($2<15200000) $2=15200000}; {if($3>15800000) $3=15800000}; {print $1,$2,$3,\"Alu\",0,\".\",$2,$3,\"204,102,0\"}' | uniq | gzip > Alu.bed.gz\n",
    "\n",
    "find . -type f -name '*.bed.gz' -exec zcat {} \\; >> allGenomeLabel.bed\n",
    "zcat exon.bed.gz intron.bed.gz intergenic.bed.gz > genomicLabel.bed\n",
    "zcat coding.bed.gz noncoding.bed.gz > codingLabel.bed\n",
    "zcat coding.bed.gz noncoding.bed.gz intron.bed.gz intergenic.bed.gz > genomicCodingLabel.bed\n",
    "zcat promoter.bed.gz > promoter.bed\n",
    "zcat LINE.bed.gz SINE.bed.gz Alu.bed.gz > repetitiveLabel.bed\n",
    "\n",
    "find . -type f -name '*.bed.gz' -exec mv '{}' ./genomeLabel/ \\;\n",
    "cd genomeLabel;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Feature|Colour\n",
    ":--:|:--:\n",
    "Exon|Green\n",
    "Intron|Red\n",
    "Intergenic|Blue\n",
    "Coding|Dark-green\n",
    "Non-coding|Light-green\n",
    "Promoter|Purple\n",
    "LINE|Yellow\n",
    "SINE|Orange\n",
    "Alu|Dark-orange"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [Steps to write and execute a script](http://omgenomics.com/writing-bash-script/):\n",
    "> - \"nano **myscript**\" opens the file editor to create/edit the script \"myscript\"  \n",
    "> - \"chmod +x **myscript**\" makes the script executable  \n",
    "> - \"./**myscript**\" or, more generally, \"*path_to_script*/**myscript**\" runs the script  "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "#!/usr/bin/env perl\n",
    "#\n",
    "# Use script to create exonic, intronic, and intergenic regions from a GTF file\n",
    "# You must provide a \"chrom.size\" file, which is simply a flat file that specifies the size of chromosomes\n",
    "# For an example, take a look at hg19.genome, which was created using the UCSC Genome Browser's MySQL database by running:\n",
    "#\n",
    "#    mysql --user=genome --host=genome-mysql.cse.ucsc.edu -A -e \\\n",
    "#    \"select chrom, size from hg19.chromInfo\" > hg19.genome\n",
    "#\n",
    "\n",
    "use strict;\n",
    "use warnings;\n",
    "use File::Which;\n",
    "use File::Basename;\n",
    "\n",
    "my $bedtools = which('bedtools');\n",
    "\n",
    "if (!$bedtools){\n",
    "   print STDERR \"bedtools was not found in your path:\\n\\n$ENV{PATH}\\n\\nPlease install bedtools and add it to your path:\\n\\n\";\n",
    "   print STDERR \"git clone https://github.com/arq5x/bedtools2.git\\ncd bedtools2\\nmake clean\\nmake all\\n\\n\";\n",
    "   exit(1);\n",
    "}\n",
    "\n",
    "my $usage = \"Usage: $0 <infile.gtf> <chrom.size>\\n\";\n",
    "my $infile = shift or die $usage;\n",
    "my $genome = shift or die $usage;\n",
    "\n",
    "if ($infile !~ /\\.gtf/){\n",
    "   print STDERR \"Please provide a GTF file\\n\";\n",
    "   exit(1);\n",
    "}\n",
    "\n",
    "my $exon_file = \"exon.bed.gz\";\n",
    "my $intergenic_file = \"intergenic.bed.gz\";\n",
    "my $intron_file = \"intron.bed.gz\";\n",
    "my $coding_file = \"coding.bed.gz\";\n",
    "my $noncoding_file = \"noncoding.bed.gz\";\n",
    "\n",
    "if (!-e $exon_file){\n",
    "   warn \"Creating exonic regions\\n\";\n",
    "   my $command = \"cat $infile | awk -F \\\"\\\\t\\\" 'BEGIN{OFS=\\\"\\\\t\\\";} (\\$1==\\\"chrX\\\" && ((\\$5 >= 15200000 && \\$5 <= 15800000) || (\\$4 >= 15200000 && \\$4 <= 15800000))) && (\\$3==\\\"exon\\\") {print \\$1,\\$4,\\$5}' | bedtools sort | bedtools merge -i stdin | awk 'BEGIN{OFS=\\\"\\\\t\\\";}; {if(\\$2<15200000) \\$2=15200000}; {if(\\$3>15800000) \\$3=15800000}; {print \\$1,\\$2,\\$3,\\\"exon\\\",0,\\\".\\\",\\$2.\\$3,\\\"0,255,0\\\"}' | uniq | gzip > $exon_file\";\n",
    "   system($command);\n",
    "} else {\n",
    "   warn \"$exon_file already exists; skipping exon step\\n\";\n",
    "}\n",
    "\n",
    "if (!-e $intron_file){\n",
    "   warn \"Creating intronic regions\\n\";\n",
    "   my $command = \"cat $infile | awk -F \\\"\\\\t\\\" 'BEGIN{OFS=\\\"\\\\t\\\";} (\\$1==\\\"chrX\\\" && ((\\$5 >= 15200000 && \\$5 <= 15800000) || (\\$4 >= 15200000 && \\$4 <= 15800000))) && (\\$3==\\\"transcript\\\") {print \\$1,\\$4,\\$5}' | bedtools sort | bedtools merge -i stdin | bedtools subtract -a stdin -b $exon_file | awk 'BEGIN{OFS=\\\"\\\\t\\\";}; {if(\\$2<15200000) \\$2=15200000}; {if(\\$3>15800000) \\$3=15800000}; {print \\$1,\\$2,\\$3,\\\"intron\\\",0,\\\".\\\",\\$2,\\$3,\\\"255,0,0\\\"}' | gzip > $intron_file\";\n",
    "   system($command);\n",
    "} else {\n",
    "   warn \"$intron_file already exists; skipping intron step\\n\";\n",
    "}\n",
    "\n",
    "if (!-e $intergenic_file){\n",
    "   warn \"Creating intergenic regions\\n\";\n",
    "   my $command = \"cat $infile | awk -F \\\"\\\\t\\\" 'BEGIN{OFS=\\\"\\\\t\\\";} (\\$1==\\\"chrX\\\" && ((\\$5 >= 15200000 && \\$5 <= 15800000) || (\\$4 >= 15200000 && \\$4 <= 15800000))) && (\\$3==\\\"transcript\\\") {print \\$1,\\$4,\\$5}' | bedtools sort | bedtools complement -i stdin -g $genome | awk 'BEGIN{OFS=\\\"\\\\t\\\"}; {if(\\$2<15200000) \\$2=15200000}; {if(\\$3>15800000) \\$3=15800000}; \\$3>=15200000 && \\$2 <=15800000' | awk 'BEGIN{OFS=\\\"\\\\t\\\";} {print \\$1,\\$2,\\$3,\\\"intergenic\\\",0,\\\".\\\",\\$2,\\$3,\\\"0,0,255\\\"}' | gzip > $intergenic_file\";\n",
    "   system($command);\n",
    "} else {\n",
    "   warn \"$intergenic_file already exists; skipping intergenic step\\n\";\n",
    "}\n",
    "\n",
    "if (!-e $coding_file){\n",
    "   warn \"Creating coding exonic regions\\n\";\n",
    "   my $command = \"cat $infile | awk -F \\\"\\\\t\\\" 'BEGIN{OFS=\\\"\\\\t\\\";} (\\$1==\\\"chrX\\\" && ((\\$5 >= 15200000 && \\$5 <= 15800000) || (\\$4 >= 15200000 && \\$4 <= 15800000))) && (\\$3==\\\"CDS\\\" || \\$3==\\\"start_codon\\\" || \\$3==\\\"stop_codon\\\")' |  bedtools sort | bedtools merge -i stdin | awk 'BEGIN{OFS=\\\"\\\\t\\\";}; {if(\\$2<15200000) \\$2=15200000}; {if(\\$3>15800000) \\$3=15800000}; {print \\$1,\\$2,\\$3,\\\"coding\\\",0,\\\".\\\",\\$2,\\$3,\\\"0,102,0\\\"}' | uniq | gzip > $coding_file\";\n",
    "   system($command);\n",
    "} else {\n",
    "   warn \"$coding_file already exists; skipping coding step\\n\";\n",
    "}\n",
    "\n",
    "if (!-e $noncoding_file){\n",
    "   warn \"Creating non-coding exonic regions\\n\";\n",
    "   my $command = \"cat $infile | awk -F \\\"\\\\t\\\" 'BEGIN{OFS=\\\"\\\\t\\\";} (\\$1==\\\"chrX\\\" && ((\\$5 >= 15200000 && \\$5 <= 15800000) || (\\$4 >= 15200000 && \\$4 <= 15800000))) && ((\\$3~/UTR/) || (\\$9~/NR/ && \\$3==\\\"exon\\\")) {print \\$1,\\$4,\\$5}' | bedtools sort | bedtools subtract -a stdin -b $coding_file | sort -k1,1 -k2,2n | bedtools merge -i stdin | awk 'BEGIN{OFS=\\\"\\\\t\\\";}; {if(\\$2<15200000) \\$2=15200000}; {if(\\$3>15800000) \\$3=15800000}; {print \\$1,\\$2,\\$3,\\\"noncoding\\\",0,\\\".\\\",\\$2,\\$3,\\\"153,255,153\\\"}' | uniq | gzip > $noncoding_file\";\n",
    "   system($command);\n",
    "} else {\n",
    "   warn \"$noncoding_file already exists; skipping non-coding step\\n\";\n",
    "}\n",
    "\n",
    "\n",
    "if (-e $exon_file && -e $intron_file && -e $intergenic_file){\n",
    "   my ($exon_average, $exon_coverage) = stats($exon_file);\n",
    "   my ($intergenic_average, $intergenic_coverage) = stats($intergenic_file);\n",
    "   my ($intron_average, $intron_coverage) = stats($intron_file);\n",
    "\n",
    "   my $total = $exon_coverage + $intergenic_coverage + $intron_coverage;\n",
    "\n",
    "   print \"\\nCoverage summary per region (percentage)\\n\\n\";\n",
    "   printf \"Exon: %.2f\\n\", $exon_coverage*100/$total;\n",
    "   printf \"Intron: %.2f\\n\", $intron_coverage*100/$total;\n",
    "   printf \"Intergenic: %.2f\\n\", $intergenic_coverage*100/$total;\n",
    "\n",
    "   print \"\\nAverage length per region (bp)\\n\\n\";\n",
    "   print \"Exon: $exon_average\\n\";\n",
    "   print \"Intron: $intron_average\\n\";\n",
    "   print \"Intergenic: $intergenic_average\\n\";\n",
    "   print \"\\n\";\n",
    "}\n",
    "\n",
    "sub stats {\n",
    "\n",
    "   my ($infile) = @_;\n",
    "   my $coverage = 0;\n",
    "   my $total = 0;\n",
    "   my $average = 0;\n",
    "\n",
    "   open(IN, '-|' ,\"zcat $infile\") || die \"Could not open $infile: $!\\n\";\n",
    "   while(<IN>){\n",
    "      chomp;\n",
    "      ++$total;\n",
    "      my ($chr, $start, $end) = split(/\\t/);\n",
    "      my $c = $end - $start;\n",
    "      $coverage += $c;\n",
    "   }\n",
    "   close(IN);\n",
    "\n",
    "   $average = sprintf(\"%.2f\", $coverage / $total);\n",
    "\n",
    "   return($average, $coverage);\n",
    "}\n",
    "\n",
    "exit(0);"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "#!/usr/bin/env perl\n",
    "\n",
    "use strict;\n",
    "use warnings;\n",
    "\n",
    "my $exon_file = \"exon.bed.gz\";\n",
    "my $intergenic_file = \"intergenic.bed.gz\";\n",
    "my $intron_file = \"intron.bed.gz\";\n",
    "my $coding_file = \"coding.bed.gz\";\n",
    "my $noncoding_file = \"noncoding.bed.gz\";\n",
    "my $promoter_file = \"promoter.bed.gz\";\n",
    "my $LINE_file =\"LINE.bed.gz\";\n",
    "my $SINE_file = \"SINE.bed.gz\";\n",
    "my $Alu_file = \"Alu.bed.gz\";\n",
    "\n",
    "my $exon_coverage = coverage($exon_file);\n",
    "my $intergenic_coverage = coverage($intergenic_file);\n",
    "my $intron_coverage = coverage($intron_file);\n",
    "my $coding_coverage = coverage($coding_file);\n",
    "my $noncoding_coverage = coverage($noncoding_file);\n",
    "my $promoter_coverage = coverage($promoter_file);\n",
    "my $LINE_coverage = coverage($LINE_file);\n",
    "my $SINE_coverage = coverage($SINE_file);\n",
    "my $Alu_coverage = coverage($Alu_file);\n",
    "\n",
    "my $total = $exon_coverage + $intergenic_coverage + $intron_coverage;\n",
    "\n",
    "printf \"Exons: %.2f\\n\", $exon_coverage*100/$total;\n",
    "printf \"Intergenic regions: %.2f\\n\", $intergenic_coverage*100/$total;\n",
    "printf \"Introns: %.2f\\n\", $intron_coverage*100/$total;\n",
    "printf \"Coding exons: %.2f\\n\", $coding_coverage*100/$total;\n",
    "printf \"Noncoding exons: %.2f\\n\", $noncoding_coverage*100/$total;\n",
    "printf \"Promoters: %.2f\\n\", $promoter_coverage*100/$total;\n",
    "printf \"LINE elements: %.2f\\n\", $LINE_coverage*100/$total;\n",
    "printf \"Non-Alu SINE elements: %.2f\\n\", $SINE_coverage*100/$total;\n",
    "printf \"Alu elements: %.2f\\n\", $Alu_coverage*100/$total;\n",
    "printf \"Repetitive elements: %.2f\\n\", ($LINE_coverage+$SINE_coverage+$Alu_coverage)*100/$total;\n",
    "\n",
    "sub coverage {\n",
    "   my ($infile) = @_;\n",
    "   my $coverage = 0;\n",
    "   open(IN,'-|',\"zcat $infile\") || die \"Could not open $infile: $!\\n\";\n",
    "   while(<IN>){\n",
    "      chomp;\n",
    "      my ($chr, $start, $end) = split(/\\t/);\n",
    "      my $c = $end - $start;\n",
    "      $coverage += $c;\n",
    "   }\n",
    "   close(IN);\n",
    "   return($coverage);\n",
    "}\n",
    "\n",
    "exit(0);"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
